pkg/services/ngalert/notifier/encryption.go

```diff
@@ -0,0 +1,108 @@
+package notifier
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+
+	"github.com/grafana/grafana/pkg/services/ngalert/api/tooling/definitions"
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/services/ngalert/store"
+	"github.com/grafana/grafana/pkg/services/secrets"
+)
+
+type Encryption interface {
+	LoadSecureSettings(ctx context.Context, orgId int64, receivers []*definitions.PostableApiReceiver) error
+	getDecryptedSecret(r *definitions.PostableGrafanaReceiver, key string) (string, error)
```

Comment:
It needs access to a `secrets.Service`, which the current caller doesn't have access to (aside through a `Crypto` instance). So, this can't be extracted out to a plain `func`, meaning we'd need to inject a concrete `encryption` into `MultiOrgAlertmanager`.

Given we are creating this abstraction `Crypto`, it felt wrong to me to have another component completely circumvent the abstraction and use the concrete underlying type directly - even if the consumer happens to be in the same package. Especially since a language feature seems to exist for precisely this situation (un-exported methods on exported interfaces).

Reply:
> Given we are creating this abstraction Crypto, it felt wrong to me to have another component completely circumvent the abstraction and use the concrete underlying type directly

At the current implementation, this would only make sense if you choose to use delegation. If you choose not to use delegation then I'm fine with the interface having a private method.

I think it's perfectly fine to go around an interface and use the concrete implementation **as long as we're within the same package** - I do understand however this is something worth examining on a case-by-case basis. There's a lot of value in using your interfaces within the same package to validate that you've got the correct abstraction.